### PR TITLE
osx external post module: terminal savedstate

### LINKED ISSUE #12440: OSX Terminal savedState File Post Module
CrowdStrike and FireEye have released [tool(s)](https://github.com/CrowdStrike/automactc/blob/master/modules/mod_terminalstate_v100.py) to handle terminal savedState files on OSX.

FireEye talked about an OSX IR where they were able to pull these files and reconstruct whole screens from the terminal.  Great for pulling back history of commands, even if someone unsets the history file.

Pretty simple, `AES-128-CBC`, the key is in one file, then you can decrypt the data file.

I'd like to take a stab at this if no one else has done it or is thinking about it.  I think it would be a really cool addition to the framework, being able to potentially gather info.

### PR DESCRIPTION
**Pull request by h00die**
Fixes #12440

This PR is a post module for OSX that enumerates, and decrypts terminal savedstate files.  These files are a complete input/output of terminals, and will hopefully include things like `cat ~/.ssh/id_rsa` or other useful tidbits.

This is an external PYTHON 3 POST module.  With that being said, that isn't a thing, so this will be a draft until a external module shim is created for POST modules.    It is based off the working https://github.com/h00die/osx_window_state but it hasn't been tested in ANY way since there isn't a shim.

Credit where credit is due, @williballenthin did all the research, coding, and responded to all my questions (physical and virtual).  So big +1 to him being awesome.

### CODE REVIEW

File: modules/post/osx/gather/terminal_savedstate.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        The files themselves contain an exact copy of what was sent
+        to and from the terminal, which may include sensitive information.
+    ''',
+    'authors': [
+        'h00die', # msf module
+        'Willi Ballenthin <willi.ballenthin@gmail.com>', # PoC
+        'kshitij Kumar <kshitij.kumar@crowdstrike.com>' # PoC
+    ],
+    'date': '2019-10-01',
+    'license': '<license>',
```
Comment (h00die) [inline modules/post/osx/gather/terminal_savedstate.py]: remove/fix

**Comment by wvu-r7** [comment]
I guess it's my turn! https://github.com/rapid7/metasploit-framework/issues/12440#issuecomment-554087107

**Comment by h00die** [comment]
of note, `read_file`, `exec` and `store_loot` were the metasploit post/report functions I assumed would be available.

**Comment by acammack-r7** [comment]
Looks like fun! A quick note: 'lib/msf/core/post/osx/bplist.py' will need to move, probably somewhere under `lib/msf/core/modules/external/python`.

**Comment by h00die** [comment]
@wvu-r7 got any time coming up to look at this one?

**Comment by wvu-r7** [comment]
Yeah, I can soon. CTF is over.

File: modules/post/osx/gather/terminal_savedstate.py (hunk 1 above)
Comment (wvu-r7) [reply modules/post/osx/gather/terminal_savedstate.py:42]: This broke loading. I changed it to `BSD_LICENSE` as per https://github.com/CrowdStrike/automactc/blob/master/LICENSE. I didn't see a license for https://gist.github.com/williballenthin/ab23abd5eec5bf5a272bfcfb2342ec04.

**Comment by h00die** [comment]
I had chatted with will, he said no license free to do whatever with the code. I think one of the url references may link to the got chat for it

**Comment by wvu-r7** [comment]
I assume you haven't been able to test this external module yet due to the lack of a post module template. I've fixed that, but there are a bunch of other problems with the module, probably due to the inability to test. I'll push up a commit or two soon.

**Comment by h00die** [comment]
yup exactly, i got it working in regular python, slapped it in the middle of the template, and couldn't load it.  Soon as you have a working post module template, lemme know and i'll finish it all up

**Comment by wvu-r7** [comment]
Ah, if that's the case, I'll abandon my other changes. I'll let you handle it, but I'll comment what needs to change.

File: modules/post/osx/gather/terminal_savedstate.py (hunk 1 above)
Comment (h00die) [reply modules/post/osx/gather/terminal_savedstate.py:42]: https://gist.github.com/williballenthin/994db929b1448fdf73baf91207129dec#file-macos_savedstate-py-L5

**Comment by wvu-r7** [comment]
I had to proceed with my changes. I added iTerm2 support, since that's the only way I was able to test the module. There is no `store_loot`, so you'll want to figure something out for that. I'm dumping to `stdout` at the moment, and that sucks.

**Comment by wvu-r7** [comment]
```
msf5 post(osx/gather/terminal_savedstate) > options

Module options (post/osx/gather/terminal_savedstate):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   SESSION  -1               yes       The session to run this module on.
   USER     wvu              yes       Which user's terminal saved state files to pull

msf5 post(osx/gather/terminal_savedstate) > run

[!] SESSION may not be compatible with this module.
[*] Enumerating files for wvu
[-] [Errno 2] No such file or directory: '/Users/wvu/Library/Saved Application State/com.apple.Terminal.savedState/windows.plist'
[*] windows.plist loaded from /Users/wvu/Library/Saved Application State/com.googlecode.iterm2.savedState/windows.plist
[*] data.data loaded from /Users/wvu/Library/Saved Application State/com.googlecode.iterm2.savedState/data.data
[*] found window: 1
[*] found archived class: _NSWindow
[*] skipping window, no title
[*] found window: 2
[*] found archived class: _NSWindow
[*] Window 1 Title: 1. ruby
[*] Terminal output 1 for user wvu:
[snip]
Day changed to 01 Apr 202011:47 < busterb> Howdy all.[snip]w v u @ k h a r a k : ~ $  ncat -e /bin/sh 127.0.0.1 4444
[snip]
[*] found window: 4294967295
[*] found archived class: _NSApplication
[*] skipping window, no title
[*] found window: 2
[*] found archived class: _NSWindow
[*] Window 3 Title: 1. ruby
[*] Terminal output 3 for user wvu:
[snip]
[*] Post module execution completed
msf5 post(osx/gather/terminal_savedstate) >
```

**Comment by h00die** [comment]
well that's looking positive! Let me know when youre done so I can start mucking.

**Comment by wvu-r7** [comment]
I'm done. Feel free to tear into it. I just wanted to get it in a working state.

**Comment by wvu-r7** [comment]
@h00die: Did you make any progress on this?

**Comment by h00die** [comment]
Yes and no, sent on slack.
Basically it works for you because those command are executing locally on your osx box, not on the remote target. I didn't see a `cmd_exec` equivalent unless I missed something, and thus no way to run remotely

**Comment by wvu-r7** [comment]
Oh, I logged off Metasploit Slack. Sorry.

**Comment by wvu-r7** [comment]
Did you intend to upload the script to the target or fetch the files that were needed? Also, lmao, I totally executed code against myself instead of running it through the session. Major derp.

**Comment by h00die** [comment]
fetch files that were needed.  Better to run locally IMO, more controlled/stable env than a target may have.

So looks like we meed:

`ls`
1. https://github.com/rapid7/metasploit-framework/pull/12688/files#diff-d007c74b8b367379f11a1e97bd32bed7R170

`file read`

1. https://github.com/rapid7/metasploit-framework/pull/12688/files#diff-d007c74b8b367379f11a1e97bd32bed7R183
2. https://github.com/rapid7/metasploit-framework/pull/12688/files#diff-d007c74b8b367379f11a1e97bd32bed7R187

**Comment by wvu-r7** [comment]
Since we still don't have `store_loot`, we'll have to do the heavy lifting ourselves, unless you want to implement it for the shim.

**Comment by h00die** [comment]
I'm leaning towards just printing the output to the screen with big separators.bthisbis already implementation a bunch of new stuff as it is

**Comment by wvu-r7** [comment]
So are you going to convert this over to use the session, or should I?

**Comment by h00die** [comment]
Can you access the session already or does that plumbing need to be built?

**Comment by wvu-r7** [comment]
I don't remember. I'll take a look. This might be way more involved than we expected, or it might be easy.

**Comment by wvu-r7** [comment]
I haven't found a good idea for `cmd_exec` yet. This might be tough.

**Comment by wvu-r7** [comment]
Maybe some sort of callback to the template. I'll see tomorrow if that pans out.

**Comment by wvu-r7** [comment]
I've rebased this branch against `master` to get rid of the unnecessary merges of `master`, since you're aware we're collaborating.

**Comment by h00die** [comment]
No problems here!

**Comment by wvu-r7** [comment]
Atticking until I have time to implement a `cmd_exec` callback or something. Feel free to work on it if you have time.

**Comment by h00die** [comment]
I won't have time in the near future to help on MSF items unfortunately

**Comment by wvu-r7** [comment]
No worries. I hope to get back to this soon-ish, actually. I just have a bunch of other things in flight that need my immediate attention.

**Comment by wvu-r7** [comment]
All yours, if you get back to it.

**Pull request closed without merging**